crates/epaint/src/text/fonts.rs

```diff
@@ -727,13 +727,156 @@ struct GalleyCache {
 }
 
 impl GalleyCache {
+    fn layout_multiline(&mut self, fonts: &mut FontsImpl, job: LayoutJob) -> Galley {
+        let pixels_per_point = fonts.pixels_per_point;
+        let round_to_pixel =
+            move |point: f32| (point * pixels_per_point).round() / pixels_per_point;
+
+        let mut current_section = 0;
+        let mut current = 0;
+        let mut left_max_rows = job.wrap.max_rows;
+        let mut galleys = Vec::new();
+        let mut first_row_min_height = job.first_row_min_height;
+        while current != job.text.len() {
+            let end = job.text[current..]
+                .find('\n')
+                .map_or(job.text.len(), |i| i + current + 1);
+            let start = current;
+
+            let mut line_job = LayoutJob {
```

Comment:
This functions is very long.

I would love if we could break this up into parts, for instance:

```
let mut galleys = vec![];
for line_job in job.split_on_newlines() {
    …
}
Galley::concat(galleys)
```

Reply:
Should such `LayoutJob::split_on_newlines` and `Galley::concat` functions be public API or do we want to keep them private?

Reply:
I looked into splitting out these functions but it's actually non trivial, for example the splitting actually has to do the layout *while* it is doing the splitting because it has to update `LayoutJob::max_rows` as more rows are being laid out. As for `Galley::concat`, it needs access to `pixels_per_point` for correct rounding so it'd be at the very least awkward as a public API (I guess you'd need to pass in `Fonts`?) and as a private one it could accept `FontsImpl`, it also needs access to the full `LayoutJob` so it can store it inside the `Galley` and also to check `round_output_size_to_nearest_ui_point`.

I think we could do something like a `Galley::concat(fonts: &mut FontsImpl, job: LayoutJob, galleys: impl Iterator<Item = Galley>) -> Galley` and keep splitting in `GalleyCache::layout_multiline` or split it out into a method of `GalleyCache` but it has to have access to `GalleyCache` so it can't be on `LayoutJob`. Also at that point I don't know whether scattering the implementation detail of `FontsImpl` into a `Galley` method is any more clean than keeping it in `GalleyCache`.

Maybe another solution would be to actually disable this line caching if `LayoutJob::max_rows` is finite, because in that scenario there's going to be more ways to invalidate the cached lines anyways and the performance benefits will be much more situational (any edit that will cause a line-count change will cause a re-layout of all subsequent lines).